src/typecheck.rs

```diff
@@ -1155,21 +1218,23 @@ pub fn unify_(
     // t1 and t2 are roots of the type
     match (t1, t2) {
         (TypeWrapper::Concrete(s1), TypeWrapper::Concrete(s2)) => match (s1, s2) {
-            (AbsType::Dyn(), AbsType::Dyn()) => Ok(()),
+            // Subtyping rule: T <: Dyn (for any concrete type T)
+            (_, AbsType::Dyn()) => Ok(()),
             (AbsType::Num(), AbsType::Num()) => Ok(()),
             (AbsType::Bool(), AbsType::Bool()) => Ok(()),
             (AbsType::Str(), AbsType::Str()) => Ok(()),
-            (AbsType::List(tyw1), AbsType::List(tyw2)) => unify_(state, *tyw1, *tyw2),
+            (AbsType::List(tyw1), AbsType::List(tyw2)) => check_sub_(state, *tyw1, *tyw2),
             (AbsType::Sym(), AbsType::Sym()) => Ok(()),
             (AbsType::Arrow(s1s, s1t), AbsType::Arrow(s2s, s2t)) => {
-                unify_(state, (*s1s).clone(), (*s2s).clone()).map_err(|err| {
+                check_sub_(state, (*s2s).clone(), (*s1s).clone()).map_err(|err| {
```

Comment:
Yes, that's a common problem unfortunately with ownership and the error monad :( maybe in this case it can me mitigated by using more references, whenever possible. I just kept the ownership structure as it was before.